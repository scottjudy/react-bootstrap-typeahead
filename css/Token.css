/* Token */
.bootstrap-tokenizer .token {
  background-color: #e7f4ff;
  border: 0;
  border-radius: 2px;
  color: #1f8dd6;
  display: inline-block;
  line-height: 1em;
  padding: 4px 7px;
  position: relative;
}
.bootstrap-tokenizer .token:focus,
.bootstrap-tokenizer .token-removeable {
  padding-right: 21px;
}
.bootstrap-tokenizer .token-selected {
  background-color: #1f8dd6;
  color: #fff;
  outline: none;
  text-decoration: none;
}

.bootstrap-tokenizer .token {
  margin: 0 3px 3px 0;
}

.bootstrap-tokenizer .token .close-button {
  bottom: 0;
  padding: 3px 7px;
  position: absolute;
  right: 0;
  top: 0;
}
