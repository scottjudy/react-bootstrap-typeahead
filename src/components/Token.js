// @flow

import cx from 'classnames';
import React, { forwardRef } from 'react';

import type { Node } from 'react';

import ClearButton from './ClearButton';

import { withToken } from '../behaviors/token';
import { isFunction } from '../utils';

import type { OptionHandler } from '../types';

type Props = {
  active: boolean,
  children?: Node,
  className?: string,
  disabled?: boolean,
  href?: string,
  onRemove?: OptionHandler,
  readOnly?: boolean,
  tabIndex?: number,
};

const InteractiveToken = forwardRef<Props, ?HTMLElement>(
  ({ active, children, className, onRemove, tabIndex, ...props }, ref) => (
    <div
      {...props}
      className={cx(
        'rbt-token',
        'rbt-token-removeable',
        {
          'rbt-token-active': !!active,
        },
        className
      )}
      ref={ref}
      tabIndex={tabIndex || 0}>
      {children}
      <ClearButton
        className="rbt-token-remove-button"
        label="Remove"
        onClick={onRemove}
        tabIndex={-1}
      />
    </div>
  )
);

const StaticToken = ({ children, className, disabled, href }) => {
  const classnames = cx(
    'rbt-token',
    {
      'rbt-token-disabled': disabled,
    },
    className
  );

  if (href && !disabled) {
    return (
      <a className={classnames} href={href}>
        {children}
      </a>
    );
  }

  return <div className={classnames}>{children}</div>;
};

/**
 * Token
 *
 * Individual token component, generally displayed within the TokenizerInput
 * component, but can also be rendered on its own.
 */
const Token = forwardRef<Props, ?HTMLElement>((props, ref) => {
  const { disabled, onRemove, readOnly } = props;

  return !disabled && !readOnly && isFunction(onRemove) ? (
    <InteractiveToken {...props} ref={ref} />
  ) : (
    <StaticToken {...props} />
  );
});

export default withToken(Token);
