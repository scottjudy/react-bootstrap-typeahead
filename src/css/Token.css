/* Token */
.token {
  background-color: #e7f4ff;
  border: 0;
  border-radius: 2px;
  color: #1f8dd6;
  float: left;
  line-height: 1em;
  margin: 0 3px 3px 0;
  padding: 4px 21px 4px 7px;
  position: relative;
}
.token:focus,
.token-selected {
  background-color: #1f8dd6;
  color: #fff;
  outline: none;
  text-decoration: none;
}

.token-close-button {
  bottom: 0;
  padding: 3px 7px;
  position: absolute;
  right: 0;
  top: 0;
}
