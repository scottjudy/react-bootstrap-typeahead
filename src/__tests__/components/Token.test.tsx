import React from 'react';

import Token from '../../components/Token';
import { noop, prepareSnapshot, render, screen, userEvent } from '../helpers';

const option = {
  label: 'test option',
};

const ACTIVE_CLASS = 'rbt-token-active';
const DISABLED_CLASS = 'rbt-token-disabled';
const REMOVEABLE_CLASS = 'rbt-token-removeable';

const TestComponent = (props) => (
  <Token option={option} {...props}>
    This is a token
  </Token>
);

describe('<Token>', () => {
  it('renders a snapshot', () => {
    expect(
      prepareSnapshot(
        <>
          <TestComponent onRemove={noop} />
          <TestComponent disabled />
          <TestComponent href="/path/to/some/url" />
        </>
      )
    ).toMatchSnapshot();
  });

  it('renders non-removeable tokens', () => {
    render(
      <>
        <TestComponent onRemove={undefined} />
        <TestComponent disabled />
        <TestComponent readOnly />
      </>
    );

    expect(screen.queryAllByRole('button').length).toBe(0);
  });

  it('renders a removeable token', () => {
    const onRemove = jest.fn();
    render(<TestComponent onRemove={onRemove} />);

    const closeButton = screen.getByRole('button');
    const token = closeButton.parentElement;

    expect(token).toHaveClass(REMOVEABLE_CLASS);

    userEvent.click(closeButton);
    expect(onRemove).toHaveBeenCalledTimes(1);
  });

  it('makes disabled tokens non-interactive', () => {
    const { container } = render(
      <TestComponent disabled href="/path/to/some/url" />
    );

    const token = container.firstChild as Element;
    expect(token.tagName).toBe('DIV');
    expect(token).not.toHaveAttribute('href');
    expect(token).toHaveClass(DISABLED_CLASS);
  });
});

describe('<Token> event handlers', () => {
  let onBlur, onClick, onFocus, onRemove, token;

  beforeEach(() => {
    onBlur = jest.fn();
    onClick = jest.fn();
    onFocus = jest.fn();
    onRemove = jest.fn();

    const { container } = render(
      <TestComponent
        onBlur={onBlur}
        onClick={onClick}
        onFocus={onFocus}
        onRemove={onRemove}
      />
    );

    token = container.firstChild;
    expect(token).not.toHaveClass(ACTIVE_CLASS);
  });

  it('handles click events', () => {
    userEvent.click(token);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(token).toHaveClass(ACTIVE_CLASS);
  });

  it('handles focus and blur events', () => {
    token.focus();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(token).toHaveClass(ACTIVE_CLASS);

    token.blur();
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(token).not.toHaveClass(ACTIVE_CLASS);
  });

  it('handles keydown events', () => {
    userEvent.keyboard('{backspace}');

    // `onRemove` called only when token is active.
    expect(onRemove).toHaveBeenCalledTimes(0);

    token.focus();
    userEvent.keyboard('{backspace}');
    expect(onRemove).toHaveBeenCalledTimes(1);

    // Other events are ignored.
    userEvent.keyboard('{enter}');
    expect(onRemove).toHaveBeenCalledTimes(1);
  });
});
