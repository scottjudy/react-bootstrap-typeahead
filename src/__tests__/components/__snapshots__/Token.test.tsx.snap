// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Token> renders a snapshot 1`] = `
Array [
  <div
    className="rbt-token rbt-token-removeable"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    option={
      Object {
        "label": "test option",
      }
    }
    tabIndex={0}
  >
    This is a token
    <button
      aria-label="Remove"
      className="close rbt-close rbt-token-remove-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
      <span
        className="sr-only visually-hidden"
      >
        Remove
      </span>
    </button>
  </div>,
  <div
    className="rbt-token rbt-token-disabled"
  >
    This is a token
  </div>,
  <a
    className="rbt-token"
    href="/path/to/some/url"
  >
    This is a token
  </a>,
]
`;
